File: recursion/normal/phone_number_mnemonic.py
def get_phone_number_mnemonic(phone_number: str) -> list[str]:
    mnemonics: list[str] = []
    current_mnemonic = ['0'] * len(phone_number)
    get_phone_number_mnemonic_helper(0, phone_number, current_mnemonic, mnemonics)
    return mnemonics

def get_phone_number_mnemonic_helper(idx: int, phone_number: str, current_mnemonic: list[str], mnemonics: list[str]):
    if idx == len(phone_number):
        mnemonics.append(''.join(current_mnemonic))
    else:
        digit = phone_number[idx]
        letters = DIGIT_LETTERS[digit]
        for letter in letters:
            current_mnemonic[idx] = letter
            get_phone_number_mnemonic_helper(idx + 1, phone_number, current_mnemonic, mnemonics)

DIGIT_LETTERS = {
    '0': ['0'],
    '1': ['1'],
    '2': ['A', 'B', 'C'],
    '3': ['D', 'E', 'F'],
    '4': ['G', 'H', 'I'],
    '5': ['J', 'K', 'L'],
    '6': ['M', 'N', 'O'],
    '7': ['P', 'Q', 'R', 'S'],
    '8': ['T', 'U', 'V'],
    '9': ['W', 'X', 'Y', 'Z']
}

File: recursion/normal/phone_number_mnemonic.mjs
function phoneNumberMnemonic(phoneNumber) {
    const current = Array(phoneNumber.length - 1).fill("0")
    const mnemonic = [];
    getMnemonic(0, phoneNumber, current, mnemonic);
    return mnemonic
}

function getMnemonic(idx, phone, current, mnemonic) {
    if (idx === phone.length) {
        mnemonic.push(current.join(""));
    } else {
        const digit = phone[idx];
        const letters = DIGIT_LETTERS[digit];
        for (const letter of letters) {
          current[idx] = letter
          getMnemonic(idx + 1, phone, current, mnemonic)
        }
    }
}

const DIGIT_LETTERS = {
    "0": ["0"],
    "1": ["1"],
    "2": ["a", "b", "c"],
    "3": ["d", "e", "f"],
    "4": ["g", "h", "i"],
    "5": ["j", "k", "l"],
    "6": ["m", "n", "o"],
    "7": ["p", "q", "r", "s"],
    "8": ["t", "u", "v"],
    "9": ["w", "x", "y", "z"],
};

const expect = 'ZLIIL1SFSS';
const mnemonic = phoneNumberMnemonic('9544517377');

console.log(mnemonic === expect);
